### DISCUSSION TITLE
Remote configs and feature flags for A/B tests strategies

### DISCUSSION THREAD

**Discussion by GuilhermeTrevisan** [MobileNativeFoundation/discussions#201] [General]
Hi all!

I work in a team that manages a Remote Config SDK for the main project of the company. The way the SDK flows is built causes some problems for us and I'm looking for new approaches to the problem (remote config).

### How our SDK works
When the SDK is called for the first time in an app open, we get all the configs from the backend, then the response is saved in a local cache (SQLite), and we bring it to RAM memory to be accessed faster. Then the app can call a `get` method that retrieves the value of the config saved in RAM.
In the next open, the app uses the values saved from the last request until a new one is finished and the SDK gets updated values.

### Problems
In the end, all of our problems are related to the first request (that one where we get all the configs). Once the app has a lot of remote configs, this request gets very big, so we have some problems:
- overload in the backend, that have to get all the configs in the database when the SDK requests it.
- sometimes the remote configs are used with old values. Once the request for the configs is very big it can take some time, so while we don't have the newest values the SDK returns the value got in the last app open, so configs used early in the app can take two app opens to be updated for the app. It brings some problems for A/B tests and etc.
- once the backend returns all the configs, there are some of them that aren't used anymore, so we process configs that we don't need.
- high cost of data transactions (caused by the big request)
- high cost for the user's internet (same as above)

### Some things we tried
- make a selective request: the app would ask for a specific config (or some of them, like 4 or 5) and the SDK would get just it, so we would have one request per config (or per a small group), but this would be faster and smaller, also we would get just used configs. Talking to the backend teams, it would have a smaller cost. But we can't implement this approach because it brings more complexity for the app, which has to ask for a request before call de the `get` method for every config. Once the app has a lot of configs, it would be really bad to pause the company to make this adjustment.
- make the request when a `get` is called: the same idea as the last one, but the request is executed when the `get` method is called. The problem is that we get an async config's get, and building the entire app with this is really bad mainly user interfaces.
- send a delta in the request: this way the backend can return just the configs that have some update, but because of the allocation approach the backend use, it's more safety to recalculate the allocation for the user every time some config is asked, so we can't insert this logic.

### My question
Do you use another approach? 
From the point of view of Mobile, is there another way to solve the Remote Config problem? 
What problems do you have in your approach?

**Comment by pouyayarandi** [comment]
Some points may help you:
>- Do you have any versioning mechanism for your config endpoint? By versioning I mean you send an environment (your client version) and the server returns flags of your version. If yes, you don't have to get all the flags every time as some of them must be deprecated for your current version. If not, you'd better discuss it with your backend developers.
>- Do you have any procedure for deprecating flags? When you experiment with a flag, it shouldn't live more than a while. It could be an option for you to have a TTL for your flags (let's say one month).

With these tricks, you may prevent your `getFlag` endpoint from getting massive.

**Reply by thekafkaf** [reply]
@pouyayarandi I'm curious how you're handling "flag deprecation", what happens to all the users that were part of a deprecated flag? Does the feature/experience disappear or changes suddenly?

**Reply by pouyayarandi** [reply]
Those users still get the flag on their legacy versions but you as a developer don't have to care about a deprecated flag since it would not be available for the next versions. We handle it with a versioning mechanism, where different versions get different responses from the config endpoint.
